### ISSUE TITLE
Blank slots in the palette: hard to see the checks with certain gtk themes

### ISSUE THREAD

**Issue by achadwick** [mypaint/mypaint#249]
Spun off from #109 :

The chequerboard pattern used for an empty slot can be hard to see, especially with dark themes:

![A MyPaint palette panel showing some blank slots highlighted with a hand-drawn ring  surrounding them](https://cloud.githubusercontent.com/assets/61299/6553872/51d7fc9e-c64d-11e4-9908-9b47491580f5.png)

It's a bit better with light variants, but not much:

![screenshot_473](https://cloud.githubusercontent.com/assets/115782/6554897/47807eac-c6b0-11e4-9163-d19ab2d1c109.png)

We should make these checks less subtle.

**Comment by mikshaw** [comment]
makes sense to me.
what about something other than transparency grid? overlay an x or some sort of icon to clearly show there's no color in the slot.

**Comment by 0ion9** [comment]
@mikshaw Personally, I think checkerboard pattern is a good choice, since it is an established way of showing 'emptiness' in paint programs. There's just something a bit wrong with how MyPaint is drawing these (compare with GIMP or Blender, for example -- transparency checks are highly visible in these programs.)

I think we should be cautious about overlaying icons on things because if it has an icon, IME people expect to be able to click it to trigger some distinct action. But, if we are going to support automatically interpolating blank runs in the palette, then we will want to be able to drag blanks just like ordinary color entries, to create suitable runs of blanks, so no distinct action can be taken AFAICS.

**Comment by achadwick** [comment]
I think runs of blanks are covered by hightlighting and pressing add in the palette editor:

![palette editor with a cell highlighted and the pointer over the add button](https://cloud.githubusercontent.com/assets/61299/6555905/7b71c1e8-c661-11e4-902b-569125555fca.png)

Though maybe a `|<---------->|` run marker could be deployed at times when a context menu might show an interpolate option.

(Sorry, bit off-topic. This is more a #109 thing)

**Comment by mikshaw** [comment]
@0ion9 
Yeah, that makes sense. My comment was more a response to "It's a bit better with light variants, but not much", which i took as meaning he was considering using something other than the grid pattern.

Personally that light grid looks perfectly fine to me, and curious if there is a problem using it with the dark theme.

**Comment by achadwick** [comment]
@mikshaw It's derived from the theme background colour, lightened and darkened a bit. But I bet the range is being clipped for the dark theme above. I suspect trying to match the bg colour is still best for the UI, but try to keep a the full range of variance.

Code is in `_palette_render()` in `gui/colors/paletteview.py` if anyone wants to play with the range.

**Comment by 0ion9** [comment]
Setting HIGHLIGHT_DLUMA to 0.25 is a start.  This works well with a light theme and I'm just installing some dark themes to test it with.

Personally, before looking at the code, I was thinking: "we should have 128 difference between the light and dark check colors. Get the bgcolor and clamp it to the range (64, 192). Then take dark color as bgcolor - 64 and light color as bgcolor + 64". Which I could still try, it just needs some adjustment to deal with `HCYColor`s.

**Comment by achadwick** [comment]
±64 looks way too high on my screen. The little squares start to dominate too much, an the palette isn't really about those. What do you think about the current master as of 2c4e75d ?

![Adwaita dark variant with recent changes](https://cloud.githubusercontent.com/assets/61299/6567685/c2081708-c6c3-11e4-9789-69f01bd86d68.png) ![Adwaita dark variant with recent changes](https://cloud.githubusercontent.com/assets/61299/6567700/f1ececdc-c6c3-11e4-8e02-c3d2d318e2ab.png)

that only has the two fixes above and no change to HIGHLIGHT_DLUMA?

I might be persuaded to bump DLUMA up to 0.1 (from its current 0.05), but I really don't like the idea of going above that. Keep the empty ones subtle, and the colour chips bold and bright :)

**Comment by achadwick** [comment]
\*turns off redshift and goes back to a recent HQ calibration with a D65 white point and ultra!bright! backlight\*

Yeah, it's OK like that too, especially the light one.

(Ow, my eyes)

**Comment by 0ion9** [comment]
@achadwick : heh, it seems we had different aims in mind, I was aiming to make them quite visible, since they would only really be used a) when adding new colors by selection, or b) when adding new colors by interpolation. I'm not sure why anyone would actually want to leave blanks around for much time.

The +- 64 idea I edited out, as the HIGHLIGHT_DLUMA adjustment seemed to work okay. Though maybe it's equivalent -- I thought HCY Y was linear though.

Your dark theme screenshot looks tolerable now. The light one seems virtually invisible -- actually, the light one IS literally completely invisible. Try blowing it up as large as you can, no sub-squares show up.

**Comment by achadwick** [comment]
@0ion9 can you inline the Gnome-Cupertino results please? Looks like a mid-grey one, and I just want to make sure that's OK too.

Hopefully non-insulting question! Do you know whether your display's calibrated and/or properly adjusted?
As I understand it, assuming you have approximately-average vision, `#ffffff` ought really be distinguishable from `#e6e6e6` - and if it isn't then your display equipment may need tweaking. If you have an LCD monitor, could you run through the tests at http://www.lagom.nl/lcd-test/ please, in particular the "Contrast" and "White saturation" ones, and maybe tweak your display settings? Distinctions shouldn't all wash out at the top of the brightness ramp.

(If I'm wrong, tell me that I'm wrong and why please! :beginner:
I'd be happy to bump up the contrast some more to make the slots more obvious with a wider range of equipment, both biological and technological! Using a `HIGHLIGHT_DLUMA` of ±0.1 should give the popular "web-safe" increment of +0x33 (+51) for grey themes. But those are *big* steps.)

**Comment by 0ion9** [comment]
@achadwick : Yes, it is calibrated: I calibrate all displays I can get my hands on using that site, as far as possible :)

That said, I could never get the white saturation to show all slots. Currently, 200 is visible and that's it! I think that's gotten worse over time, I had up to 248 visible last time, IIRC. No amount of brightness/saturation/gamma adjustment makes those squares visible, they are just the same value (not literally -- they eyedrop as the correct values -- , but visually)

It's also possible it could be the video card. I'll try hooking the monitor up to a laptop later and see if I get different results.

Gnome-Cupertino, which has pretenses of being medium grey:

![screenshot_479](https://cloud.githubusercontent.com/assets/115782/6575635/83bbc9f0-c780-11e4-9df6-9e516bf1acc4.png)

and MediterraneanNight-mod2, part of the Oxide XFCE theme, which actually *is* medium grey:

![screenshot_478](https://cloud.githubusercontent.com/assets/115782/6575652/9e0343e2-c780-11e4-8c74-217c98917434.png)

(In case it's unclear: I can't see the checks on Gnome-Cupertino, but I can easily see them on MediterraneanNight-mod2)

**Comment by mikshaw** [comment]
latest commit looks good here, both dark and light themes (whatever the debian defaults are), on my external monitor. It's quite a bit harder to see in light theme on the laptop monitor. Not something that affects me personally, since i use dark theme and only paint using the external monitor, but it might be a problem for some (calibrating a laptop monitor isn't something many people actually do, including myself)

**Comment by 0ion9** [comment]
Yeah, calibration of laptop monitors ranges between 'minimal' and 'not supported at all'. The only way to get around that really is to use xrandr's gamma settings to compensate somewhat (and hope that your particular video hardware supports that properly)

**Comment by 0ion9** [comment]
FWIW after switching to a DVI cable, I've confirmed that my white saturation is fixed (Previously, DAC problems meant that 226 intensity was the highest displayable) and calibration is decent, so I'll reassess:

* Gnome-Cupertino and MediterraneanNight-mod2 both have good visibility. 
* The image that @achadwick posted in relation to 2c4e75d has moderate visibility on the dark theme and good visibility on the light theme.
* The images in the OP have pathetic visibility on dark themes and okay visibility on the light theme.
* The images I posted with DLUMA = 0.25 have excessive visibility on light themes and good visibility on dark themes.

So user/hardware error, basically. I think that basically places us all in agreement. Perhaps the DLUMA could be raised by 30%  (to 0.0665) to improve visibility on dark themes.
